### PR TITLE
8261812: C2 compilation fails with assert(!had_error) failed: bad dominance

### PR DESCRIPTION
**Pull request by rwestrel**
In the testByte() test case, byteField is assigned int values that
don't fit in a byte.

obj.byteField = (byte)(1 << i);

At parse time, c2 expands this to (in pseudo code):

int v = (1 << i);
obj.byteField = (v << 24) >> 24;

next:

StoreBNode::Ideal() runs and StoreNode::Ideal_sign_extended_input()
causes this to become:

obj.byteField = (1 << i);

i is then constant folded to 9:

obj.byteField = 512;

obj doesn't escape and is scalarized. The value of the byteField needs
to be recorded in the debug info. Because there are 2 such a store, a
Phi is created:

(Phi 512 1024)

The Phi is created with type TypeInt::BYTE. Because that doesn't
overlap with the input values, the Phi is transformed to top. The test
case then triggers a deoptimization and a crash occurs in the
deoptimization code.

The failure of the bug report is not a crash during deoptimization but
a c2 crash. The root cause is the same but in that case a chain of Phi
is involved and when one becomes top, it causes an uncommon trap to
have an input that doesn't dominate the uncommon trap.

I noticed LoadBNode::Ideal() re-inserts the 2 shifts when it returns a
value from a Store. I propose the same fix here in the allocation
elimination code.

Same problem applies to short and char. They are covered by the fix. I
also included boolean even though I don't see how to trigger the
failure with it.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261812](https://bugs.openjdk.java.net/browse/JDK-8261812): C2 compilation fails with assert(!had_error) failed: bad dominance

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 7ec1fab0998682a8cbae6fd418b311a1adfaa83a
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2774/head:pull/2774`
`$ git checkout pull/2774`

### CODE REVIEW

File: src/hotspot/share/opto/macro.cpp [hunk 1]
```diff
@@ -450,6 +450,9 @@ Node *PhaseMacroExpand::value_from_mem_phi(Node *mem, BasicType ft, const Type *
         Node* n = val->in(MemNode::ValueIn);
         BarrierSetC2* bs = BarrierSet::barrier_set()->barrier_set_c2();
         n = bs->step_over_gc_barrier(n);
+        if (ft == T_SHORT || ft == T_BYTE || ft == T_CHAR || ft == T_BOOLEAN) {
```
Comment (vnkozlov) [inline src/hotspot/share/opto/macro.cpp:453]: Use `is_subword_type(bt)` defined in `globalDefinitions.hpp`

**Review by vnkozlov — CHANGES_REQUESTED** [review]
Otherwise looks good.

File: src/hotspot/share/opto/macro.cpp (hunk 1 above)
Comment (rwestrel) [reply src/hotspot/share/opto/macro.cpp]: Thanks for the review. I pushed a commit with your suggestion.

**Review by vnkozlov — APPROVED** [review]
Good.

File: src/hotspot/share/opto/compile.cpp [hunk 2]
```diff
@@ -4793,3 +4793,27 @@ void Compile::igv_print_method_to_network(const char* phase_name) {
...
+  if (bt == T_BYTE) {
+    result = phase->transform(new LShiftINode(value, phase->intcon(24)));
+    result = new RShiftINode(result, phase->intcon(24));
+  } else if (bt == T_BOOLEAN) {
+    result = new AndINode(value, phase->intcon(0xFF));
+  } else if (bt == T_CHAR) {
+    result = new AndINode(value,phase->intcon(0xFFFF));
+  } else {
+    assert(bt == T_SHORT, "unexpected narrow type");
+    result = phase->transform( new LShiftINode(value, phase->intcon(16)) );
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/compile.cpp:4811]: There's some excess whitespacing.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

File: src/hotspot/share/opto/compile.cpp (hunk 2 above)
Comment (rwestrel) [reply src/hotspot/share/opto/compile.cpp]: fixed. Thanks for the review.

**Review by TobiHartmann — APPROVED** [review]

**Pull request closed without merging**